openedx/core/djangoapps/content/search/api.py

```diff
@@ -0,0 +1,486 @@
+            raise TimeoutError("Timeout acquiring lock")
+
+        time.sleep(1)
+
+
+def _get_running_rebuild_index_name() -> str | None:
+    lock_id = f"lock-meilisearch-index-{INDEX_NAME}"
+
+    return cache.get(lock_id)
+
+
+def _get_meilisearch_client():
+    """
+    Get the Meiliesearch client
+    """
+    global _MEILI_CLIENT  # pylint: disable=global-statement
+
+    # Connect to Meilisearch
+    if not is_meilisearch_enabled():
+        raise RuntimeError("MEILISEARCH_ENABLED is not set - search functionality disabled.")
+
+    if _MEILI_CLIENT is not None:
+        return _MEILI_CLIENT
+
+    _MEILI_CLIENT = MeilisearchClient(settings.MEILISEARCH_URL, settings.MEILISEARCH_API_KEY)
+    try:
+        _MEILI_CLIENT.health()
+    except MeilisearchError as err:
+        _MEILI_CLIENT = None
+        raise ConnectionError("Unable to connect to Meilisearch") from err
+    return _MEILI_CLIENT
+
+
+def clear_meilisearch_client():
+    global _MEILI_CLIENT  # pylint: disable=global-statement
+
+    _MEILI_CLIENT = None
+
+
+def _get_meili_api_key_uid():
+    """
+    Helper method to get the UID of the API key we're using for Meilisearch
+    """
+    global _MEILI_API_KEY_UID  # pylint: disable=global-statement
+
+    if _MEILI_API_KEY_UID is not None:
+        return _MEILI_API_KEY_UID
+
+    _MEILI_API_KEY_UID = _get_meilisearch_client().get_key(settings.MEILISEARCH_API_KEY).uid
+
+
+def _wait_for_meili_task(info: TaskInfo) -> None:
+    """
+    Simple helper method to wait for a Meilisearch task to complete
+    """
+    client = _get_meilisearch_client()
+    current_status = client.get_task(info.task_uid)
+    while current_status.status in ("enqueued", "processing"):
+        time.sleep(1)
+        current_status = client.get_task(info.task_uid)
+    if current_status.status != "succeeded":
+        try:
+            err_reason = current_status.error['message']
+        except (TypeError, KeyError):
+            err_reason = "Unknown error"
+        raise MeilisearchError(err_reason)
+
+
+def _wait_for_meili_tasks(info_list: list[TaskInfo]) -> None:
+    """
+    Simple helper method to wait for multiple Meilisearch tasks to complete
+    """
+    while info_list:
+        info = info_list.pop()
+        _wait_for_meili_task(info)
+
+
+def _index_exists(index_name: str) -> bool:
+    """
+    Check if an index exists
+    """
+    client = _get_meilisearch_client()
+    try:
+        client.get_index(index_name)
+    except MeilisearchError as err:
+        if err.code == "index_not_found":
+            return False
+        else:
+            raise err
+    return True
+
+
+@contextmanager
+def _using_temp_index(status_cb: Callable[[str], None] | None = None) -> Generator[str, None, None]:
+    """
+    Create a new temporary Meilisearch index, populate it, then swap it to
+    become the active index.
+    """
+    def nop(_):
+        pass
+
+    if status_cb is None:
+        status_cb = nop
```

Comment:
If no `status_cb` is passed, we should log using the `log` function from line 35.

Also please document the `status_cb` parameter in the docstring; it's not super clear what it's for.

Reply:
Done: 8e64b137df3c03cd092b7b39ad93197f2123ad31